File: p2p-ws-sdk-ts/scripts/verify_vectors.ts
import fs from "node:fs"
import path from "node:path"
import { xorNoWrap } from "../src/xor.js"
import { decodeFrame } from "../src/frame.js"

function hexToBytes(s: string): Uint8Array {
  const t = s.trim()
  if (t.length % 2 !== 0) throw new Error("hex length must be even")
  const out = new Uint8Array(t.length / 2)
  for (let i = 0; i < out.length; i++) {
    const hi = parseInt(t[i * 2], 16)
    const lo = parseInt(t[i * 2 + 1], 16)
    if (Number.isNaN(hi) || Number.isNaN(lo)) throw new Error("invalid hex")
    out[i] = (hi << 4) | lo
  }
  return out
}

function bytesToHex(b: Uint8Array): string {
  return Buffer.from(b).toString("hex")
}

const root = path.resolve(import.meta.dirname, "..", "..")
const vecPath = path.resolve(root, "p2p-ws-protocol", "test-vectors", "xor_vector_001.json")
const v = JSON.parse(fs.readFileSync(vecPath, "utf-8"))

const key = hexToBytes(v.keyfile_bytes_hex)
const plain = hexToBytes(v.plain_hex)
const got = xorNoWrap(plain, key, v.offset)
const gotHex = bytesToHex(got)
const expected = String(v.cipher_hex).toLowerCase()

if (gotHex !== expected) {
  throw new Error(`mismatch: got=${gotHex} expected=${expected}`)
}

console.log("ok=1")

const framePath = path.resolve(root, "p2p-ws-protocol", "test-vectors", "frame_vector_001.json")
const fvec = JSON.parse(fs.readFileSync(framePath, "utf-8"))
const ws = hexToBytes(fvec.ws_binary_payload_hex)
const decoded = decodeFrame(ws)
if (decoded.header.length !== fvec.header.length_u32) throw new Error("frame header length mismatch")
if (decoded.header.magic !== fvec.header.magic_u16) throw new Error("frame header magic mismatch")
if (decoded.header.version !== fvec.header.version_u8) throw new Error("frame header version mismatch")
if (decoded.header.flags !== fvec.header.flags_u8) throw new Error("frame header flags mismatch")
if (bytesToHex(decoded.cipherPayload) !== String(fvec.cipher_payload_hex).toLowerCase()) throw new Error("frame payload mismatch")

console.log("ok=2")

File: p2p-ws-sdk-python/scripts/verify_vectors.py
import json
from pathlib import Path
import sys

root = Path(__file__).resolve().parents[1]
sys.path.insert(0, str(root / "src"))

from p2p_ws_sdk.xor import xor_no_wrap
from p2p_ws_sdk.frame import decode_frame


def hx(s: str) -> bytes:
    return bytes.fromhex(s.strip())


def main() -> int:
    repo = root.parent
    vec = repo / "p2p-ws-protocol" / "test-vectors" / "xor_vector_001.json"
    v = json.loads(vec.read_text(encoding="utf-8"))
    key = hx(v["keyfile_bytes_hex"])
    plain = hx(v["plain_hex"])
    got = xor_no_wrap(plain, key, int(v["offset"])).hex()
    expected = str(v["cipher_hex"]).lower()
    if got != expected:
        raise SystemExit(f"mismatch: got={got} expected={expected}")
    print("ok=1")

    frame_vec = repo / "p2p-ws-protocol" / "test-vectors" / "frame_vector_001.json"
    fv = json.loads(frame_vec.read_text(encoding="utf-8"))
    ws = hx(fv["ws_binary_payload_hex"])
    f = decode_frame(ws)
    h = fv["header"]
    if f.header.length != int(h["length_u32"]):
        raise SystemExit("frame length mismatch")
    if f.header.magic != int(h["magic_u16"]):
        raise SystemExit("frame magic mismatch")
    if f.header.version != int(h["version_u8"]):
        raise SystemExit("frame version mismatch")
    if f.header.flags != int(h["flags_u8"]):
        raise SystemExit("frame flags mismatch")
    if f.cipher_payload.hex() != str(fv["cipher_payload_hex"]).lower():
        raise SystemExit("frame payload mismatch")
    print("ok=2")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
